File: 01_07_rotate_matrix.java
class Solution {
    public void rotate(int[][] matrix) {
        
        for(int i = 0;i < matrix.length / 2;i++) {
            
            for (int j = i;j < matrix[0].length - 1 - i;j++) {
                int top = matrix[i][j];
                int right = matrix[j][matrix[0].length - 1 - i];
                int bottom = matrix[matrix[0].length - 1 - i][matrix[0].length - 1 - j];
                int left = matrix[matrix[0].length - 1 - j][i];
                
                
                matrix[i][j] = left;
                matrix[j][matrix[0].length - 1 - i] = top;
                matrix[matrix[0].length - 1 - i][matrix[0].length - 1 - j] = right;
                matrix[matrix[0].length - 1 - j][i] = bottom;
                
                System.out.println(top + " " + right + " " + bottom + " " + left);
            }
            System.out.println("New layer!");
        }
    }
    
}


class Solution {
    public void rotate(int[][] matrix) {
        for (int i = 0;i < matrix.length/2;i++) {
            for (int j = i;j < matrix.length - 1 - i;j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[matrix.length - 1 - j][i];
                matrix[matrix.length - 1 - j][i] = matrix[matrix.length - 1 - i][matrix.length - 1 - j];
                matrix[matrix.length - 1 - i][matrix.length - 1 - j] = matrix[j][matrix.length - 1 - i];
                matrix[j][matrix.length - 1 - i] = temp;
            }
        }
    }
}

File: 01_07_rotate_matrix.js
function rotate(matrix) {

    matrix.reverse()

    for(let i = 0; i < matrix.length; i++){
        for(let j = i + 1; j < matrix.length; j++) {
            let temp = matrix[i][j]
            matrix[i][j] = matrix[j][i]
            matrix[j][i] = temp
        }
    }
    return matrix
}

//reverse the array
    //[[7,8,9],[4,5,6],[1,2,3]]

    
    /*
    for(let i = 0; i < matrix.length; i++){
    [[7,8,9]]
     *for(let j = i + 1; j < matrix.length; j++)
     *let temp = matrix[i[j]]
     matrix[i[j]] = matrix[j[i]]
     matrix[j[i]] = temp
    *}

    * i=0 j=i+1
    *matrix[j]
    *[ [7,4,9], [8,5,6], [1,2,3] ]

    *[[7,4,1], [8,5,6], [9,2,3] ]

    *[[7,4,1], [8,5,2], [9,6,3] ]
    *
    */